### PR TITLE
[Merged by Bors] - feat: port Data.W.Constructions

### PR DESCRIPTION
**Pull request by int-y1**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/W/Constructions.lean [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+#align W_type.of_nat WType.ofNat
+
+/-- The isomorphism from the `WType` of the naturals to the naturals -/
+@[simp]
+def toNat : WType Natβ → ℕ
+  | WType.mk Natα.zero _ => 0
+  | WType.mk Natα.succ f => (f ()).toNat.succ
+#align W_type.to_nat WType.toNat
+
+theorem left_inv_nat : Function.LeftInverse ofNat toNat
```
Comment (jcommelin) [inline Mathlib/Data/W/Constructions.lean:69]:
```suggestion
theorem leftInverse_nat : Function.LeftInverse ofNat toNat
```

File: Mathlib/Data/W/Constructions.lean [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    rw [toNat, ofNat]
+    congr
+    ext x
+    cases x
+  | WType.mk Natα.succ f => by
+    simp only [toNat, ofNat, left_inv_nat (f ()), mk.injEq, heq_eq_eq, true_and]
+    rfl
+#align W_type.left_inv_nat WType.left_inv_nat
+
+theorem right_inv_nat : Function.RightInverse ofNat toNat
```
Comment (jcommelin) [inline Mathlib/Data/W/Constructions.lean:80]:
```suggestion
theorem rightInverse_nat : Function.RightInverse ofNat toNat
```

File: Mathlib/Data/W/Constructions.lean [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+#align W_type.of_list WType.ofList
+
+/-- The isomorphism from the `WType` construction of lists to lists -/
+@[simp]
+def toList : WType (Listβ γ) → List γ
+  | WType.mk Listα.nil _ => []
+  | WType.mk (Listα.cons hd) f => hd :: (f PUnit.unit).toList
+#align W_type.to_list WType.toList
+
+theorem left_inv_list : Function.LeftInverse (ofList γ) (toList _)
```
Comment (jcommelin) [inline Mathlib/Data/W/Constructions.lean:166]:
```suggestion
theorem leftInverse_list : Function.LeftInverse (ofList γ) (toList _)
```

File: Mathlib/Data/W/Constructions.lean [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  | WType.mk Listα.nil f => by
+    simp only [toList, ofList, mk.injEq, heq_eq_eq, true_and]
+    ext x
+    cases x
+  | WType.mk (Listα.cons x) f => by
+    simp only [ofList, left_inv_list (f PUnit.unit), mk.injEq, heq_eq_eq, true_and]
+    rfl
+#align W_type.left_inv_list WType.left_inv_list
+
+theorem right_inv_list : Function.RightInverse (ofList γ) (toList _)
```
Comment (jcommelin) [inline Mathlib/Data/W/Constructions.lean:176]:
```suggestion
theorem rightInverse_list : Function.RightInverse (ofList γ) (toList _)
```

**Review by jcommelin — COMMENTED** [review]
Seems like a good moment to make the names closer to the naming convention.

File: Mathlib/Data/W/Constructions.lean (hunk 4 above)
Comment (int-y1) [reply Mathlib/Data/W/Constructions.lean]: All done.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**